### ISSUE TITLE
Slow Submission

### ISSUE THREAD

**Issue by jmgirard** [surveydown-dev/surveydown#86]
When I click the Next button, it is very slow to go to the next page (e.g., 30-60s for only 10 items) - presumably because it is writing to the database. Are you experiencing this too? It is also somewhat unclear whether things are even working after clicking Next. Perhaps there needs to be some kind of "processing..." modal?

**Comment by jhelvy** [comment]
We recently noticed this too. One potential reason is we switched from making direct connections on each session to using a pool to manage connections to the database. We're still testing and may revert back as the pool approach is looking like it's slowing things down considerably. We had virtually no lag before making this change, so it may be an issue with how we've implemented the pool.

**Comment by jmgirard** [comment]
Ok, great. I'm glad that's not the expected user experience.

**Comment by jmgirard** [comment]
The slowdown is happening to me for the same file when rendered locally on both Windows and Mac (where Mac additionally required setting gssencmode to "disable").

**Comment by jhelvy** [comment]
Man, this is really hard to solve because I just can't reproduce any slow down. I ran your exact same survey connected to one of my supabase databases and it all went seamlessly. I even opened 5 other windows to have multiple concurrent sessions, and all the records were immediately writing to the DB. @pingfan-hu noted that one time he experienced this, but it was only immediately after creating a database. But once he ran the survey again everything was fast. So perhaps it's just a one-time thing when first starting? I still myself have not experienced anything but an immediate connection.

As a side note, the `gssencmode = 'disable'` setting isn't necessarily a mac thing but rather a setting if you're on a protected network of some kind. For example, when I'm on campus at GWU I have to use that setting otherwise I can't make the connection, or when at home and on a VPN.

**Comment by jhelvy** [comment]
Okay one more thought. I think we may have been setting up the database inefficiently in our examples. We have the sd_database function in the server code chunk, but I actually think it should be in the setup chunk. It shouldn't make a difference in terms of things working, but I believe if it's in the server the connections are less efficient (we're making a connection pool for each session, rather than for the entire app with shared connections across sessions). So you may try moving the sd_database command up to the setup chunk to see if that helps.

**Comment by jmgirard** [comment]
When I move `sd_server()` up to setup, I get the following error:

```
processing file: mwe.qmd
  |............                                        |  22% [unnamed-chunk-1]Error:
! object 'session' not found
Backtrace:
 1. surveydown::sd_server(...)
```

**Comment by pingfan-hu** [comment]
Hi @jmgirard , `sd_server()` stays as it is, but `db <- sd_database()` goes to the setup chunk.
In your case, you move this to the setup chunk:
```r
db <- sd_database(
  host   = "aws-0-us-west-1.pooler.supabase.com",
  dbname = "postgres",
  port   = "6543",
  user   = "postgres.outtobdgbkorhxfkpijc",
  table  = "mwe",
  gssencmode = "disable"
)
```

**Comment by jmgirard** [comment]
That fixed the session error, but unfortunately did not fix the speed issue.

**Comment by jmgirard** [comment]
In case it is relevant, I am seeing this in the Render output now:
```
<pool> Checked-out object deleted before being returned.
<pool> Make sure to `poolReturn()` all objects retrieved with `poolCheckout().`
```

**Comment by pingfan-hu** [comment]
Yes, I saw that, too. It means we might have raised another problem during our fixing process with the pool.

I've just tested your qmd on my machine but didn't see delay. There is a tiny delay after I click on the answers, but that's normal due to data transaction between my machine to the server.

We're getting our hands onto this issue and will update with you. Stay tuned Jeffery.

**Comment by jmgirard** [comment]
Here is a video in case it's helpful:
https://us02web.zoom.us/clips/share/QE_S80pZTVzRqZbydzC0iU_AkgHRtaU2CU1wQCjmMlONa3C6Ll3FmsqTA82xGTZ-NiqsHUcfWAT7oTAs0rImYLKH.YBS4RnX7pYJ12hlG

**Comment by pingfan-hu** [comment]
I saw the issue. Your connection with the server is super slow. We can tell from the progress bar on the top. The progress bar is an observing event, so it reveals the real connection with the server. The reason why it takes so long to get to the next page is because the server is not yet synchronized with your answer input. The next button is in queue and has to wait for the questions to be answered (aka after the progress bar is fulfilled).
This is very similar to what I'm experience when I setup my database, but the different is I only experience it once.
I saw you are using a server on the west. I recommend that you can change to an eastern server (I see you are located at KU, which should be a little closer to the east coast than west). After you change the server, render your survey twice, see if your connection could be better in the 2nd run.
In the meantime, we will get our hand onto this problem.

**Comment by jmgirard** [comment]
Switching to an Ohio server is a bit better, but still not usable. Let me know if I can help troubleshoot.

**Comment by pingfan-hu** [comment]
Haha thanks Jeffery. We will try to push an update and let you know when it's ready for a debugging.

**Comment by jhelvy** [comment]
I just watched the video and understand the issue. One more idea, though I doubt this plays any role in the problem - have you tried opening the app in your web browser? I just wonder if the RStudio browser is slow. I fully expect this has to do with the database, but just trying to head off any other possible sources of the problem.

**Comment by jmgirard** [comment]
> I just watched the video and understand the issue. One more idea, though I doubt this plays any role in the problem - have you tried opening the app in your web browser? I just wonder if the RStudio browser is slow. I fully expect this has to do with the database, but just trying to head off any other possible sources of the problem.

Unfortunately the same in Chrome.

**Comment by jhelvy** [comment]
I'm also assuming that this isn't an issue when you disconnect the database, which you can do with `ignore = TRUE` in `sd_database`. Note that we just changed that today, it used to be called `pause`. If that's the case, then this may be something with supabase that we need to address. We also plan to support other databases, so perhaps there's an alternative that is more robust.

**Comment by jmgirard** [comment]
> I'm also assuming that this isn't an issue when you disconnect the database, which you can do with `ignore = TRUE` in `sd_database`. Note that we just changed that today, it used to be called `pause`. If that's the case, then this may be something with supabase that we need to address. We also plan to support other databases, so perhaps there's an alternative that is more robust.

Correct, it is very "snappy" (i.e., no slowdown) with `ignore = TRUE`.

**Comment by pingfan-hu** [comment]
Hi @jmgirard , I've just deployed my demo survey to [this site](https://pingfan.shinyapps.io/demo/). Can you click on the link and go through the survey? Let me know if it's slow, or it's in an acceptable reaction speed.

**Comment by jmgirard** [comment]
> Hi @jmgirard , I've just deployed my demo survey to [this site](https://pingfan.shinyapps.io/demo/). Can you click on the link and go through the survey? Let me know if it's slow, or it's in an acceptable reaction speed.

It is very fast for me (nearly instant).

**Comment by pingfan-hu** [comment]
Great! Now you can also try to deploy your survey to shinyapps. See if the deployed version could be faster. You can also share your link with us so we can have a test on our end.
What I'm guessing is that the internet connection from your machine to Supabase is not strong, but I don't know the exact reason.

**Comment by jmgirard** [comment]
Ok, I can try deploying tomorrow.

**Comment by jmgirard** [comment]
I had some time free up. Here is my deployed version:
https://jmgirard.shinyapps.io/PIDMWE/
It's faster than on my localhost, but still seems slower than yours.

I also rewrote the programmatic question creation to avoid depending on tidyverse:

<details><summary>Code</summary>

````qmd
---
server: shiny
filters: [surveydown]
---

```{r}
#| context: setup
#| echo: false
#| warning: false
#| message: false

library(surveydown)

sd_setup()

items <- read.csv("pid_items.csv")

options <- c(
  "Very False or Often False" = "0",
  "Sometimes or Somewhat False" = "1",
  "Sometimes or Somewhat True" = "2",
  "Very True or Often True" = "3"
)

db <- sd_database(
  host   = "aws-0-us-east-2.pooler.supabase.com",
  dbname = "postgres",
  port   = "6543",
  user   = "postgres.cpvuzauqxqrawlhydmhb",
  table  = "mwe"
)
```

::: {#welcome .sd-page}

```{r}
tagList(
  lapply(1:10, \(i) sd_question(
    id = as.character(items[[i, "PID5"]]),
    type = "mc",
    label = items[[i, "Stem"]],
    option = options
  ))
)
```

```{r}
sd_next(next_page = 'end')
```

:::

::: {#end .sd-page}

This it the last page in the survey

:::

```{r}
#| context: server

config <- sd_config(
  admin_page = TRUE
)

sd_server(
  input   = input,
  output  = output,
  session = session,
  config  = config,
  db      = db
)
```
````

</details>

**Comment by pingfan-hu** [comment]
Yes. It's indeed a little slower than mine. I wonder if it's because one page has too many questions. Can you try to split the questions into several pages?
For example, you can try 2-3 questions for each page.
I don't know if this can help make the survey run faster. The results can also help us to debug and prevent from future survey slowdowns.

**Comment by jmgirard** [comment]
I updated it to be 3 questions on two pages. It's a bit faster now, but still too slow.

**Comment by pingfan-hu** [comment]
I noticed the 1st question of every page reacts fast. The last question reacts slow.
I believe the slowdown is caused by the number of questions on the same page. I need to test more on my own.
But it's good to see the online survey is lot faster than local.
Wait for my further progress, Jeffrey!

**Comment by jhelvy** [comment]
Okay I finally have been able to replicate the slow experience. I believe it's the internet connection speed. I'm using a hot spot from my phone and according to speedtest.net [my speeds are very slow](https://www.speedtest.net/result/16724828982): 2.9 MBPS download, 0.42 MBPS upload. Compared to my home internet where I have over 500 download and 100 upload. So I think we're limited by bandwidth here where the uploading and downloading to the database is for some reason really expensive. 

We'll need to look into why this is so slow on slow internet speeds, because the requests we're sending to the database should be otherwise very low requirements. Even with these low speeds, I can still stream a youtube video, so uploading a single row of data to a database should still be instantaneous. It may just be a limitation of the free tier with supabase (some limit on incoming connections or requests, for example), or it could be how we're implementing the database read / write in R. But this is the source of the problem. So for now, it's going to be slow on a slow internet connection, but we'll work to fix this as there's no reason it should be this slow. This has to be addressed otherwise the platform is more or less useless.

It shouldn't matter how many questions you have on any one page as all questions are already present on the app launch. They're just hidden depending on the page they're on. The only thing that matters is the total number of questions, which will just make the data size larger. But still, even like 100 questions should be super fast as I can run a 100-question survey locally with no database connection and it's all instantaneous. And I just ran a 3-question survey with a slow internet connection and it was terribly slow. So number of questions should be irrelevant.

**Comment by jmgirard** [comment]
It would surprise me if it was just client speed. I have Fiber with 371 Mbps Down and 1.53 Mbps Up.

**Comment by jhelvy** [comment]
It's definitely related to speed in some way. I just connected to a [faster wifi](https://www.speedtest.net/result/16724924650) and it's much faster on the exact same survey. Nothing else changed except my connection speed. 

But still, it's only a 10-questions survey and there is a noticeable lag. Not too bad, but still not instantaneous, which is unacceptable. I have to believe we've got something wrong in how the database connection is working. I'll experiment with some changes in the connection settings.

**Comment by jmgirard** [comment]
That just makes me wonder if perhaps the issue that you and I are experiences are different. Since mine is occurring on a T1 line.

**Comment by jhelvy** [comment]
While we're working to address the issue, one potential temporary change we could make is to adjust the data writing operation to only happen on each page turn. Currently there is a write operation that happens on any click on any question. We have it set up that way to minimize potential for data loss, but it dramatically increases the total number of write operations. If we changed it to only write on a page turn, the user wouldn't experience any lag when answering, though there may be a lag on the page turn. Less ideal, but maybe better for now. It's just a thought. Ideally we need to figure out the source of the slow down, and I'm afraid the answer may be the supabase free tier settings. I'm going to experiment with a paid tier with better speeds and if that solves it, then there may be less we can do other than look for an alternative. Our goal with supabase is that it is free and open source, but free may just equal slow.

**Comment by willking98** [comment]
I am also experiencing the speed issue described by @jmgirard but only locally. Also, unfortunately since connecting to the database my app deployed on shinyapps.io is crashing within about 2 second of loading on the page. https://willking98.shinyapps.io/demo-choice-based-conjoint-main/ . So the issue I have is very slow ran locally and crashing on shinyapps.io. When Ignore = True in db function I have no problems.

Other than the speed thing, this package is outrageously good and a fantastic effort from the development team. Really impressive work to bring things together in a completely open source and accessible way. If the speed thing is sorted I will be giving it a massive shout out at work and in my thesis!

**Comment by pingfan-hu** [comment]
> I am also experiencing the speed issue described by @jmgirard but only locally. Also, unfortunately since connecting to the database my app deployed on shinyapps.io is crashing within about 2 second of loading on the page. https://willking98.shinyapps.io/demo-choice-based-conjoint-main/ . So the issue I have is very slow ran locally and crashing on shinyapps.io. When Ignore = True in db function I have no problems.
> 
> Other than the speed thing, this package is outrageously good and a fantastic effort from the development team. Really impressive work to bring things together in a completely open source and accessible way. If the speed thing is sorted I will be giving it a massive shout out at work and in my thesis!

Hi Will! Thanks for trying out our package. I clicked on your link and yes it crashed in like 2 seconds. Is it okay for you to share your repo with us so we can take a deeper look?

**Comment by pingfan-hu** [comment]
Also, we've just pushed a new version that allows you to deploy online without connecting to the database. You can simply set ignore = TRUE and leave everything else blank, like this:
```r
db <- sd_database(
  host   = "",
  dbname = "",
  port   = "",
  user   = "",
  table  = "",
  ignore = TRUE
)
```

However, this can only create a demo-purposed survey that won't collect any data.
Here is an example: https://pingfan.shinyapps.io/survey/

You can try to install our latest version by:
```r
remotes::install_github("surveydown-dev/surveydown", force = TRUE)
```

Then you can deploy your survey again. If you can get successful this time, then it's indeed a connection problem with your Supabase database.

I'm also wondering if this survey works out fine on your local machine. Like, can you successfully connect to Supabase database while you run locally and see the data collected in your table?

**Comment by willking98** [comment]
Yes no problem I have added the code to my Github now. Link here: https://github.com/willking98/demo-choice-based-conjoint-main/tree/main 

Yep I had already deployed it to shinyapps.io prior to connecting to Supabase and it was working really quickly. I have just updated to the latest version and this has not changed anything. Yes when I run this locally I can see the data appearing in the tables on Supabase. The blue progress bar moves instantly when I select an option - in case that helps?

**Comment by pingfan-hu** [comment]
> Yes no problem I have added the code to my Github now. Link here: https://github.com/willking98/demo-choice-based-conjoint-main/tree/main
> 
> Yep I had already deployed it to shinyapps.io prior to connecting to Supabase and it was working really quickly. I have just updated to the latest version and this has not changed anything. Yes when I run this locally I can see the data appearing in the tables on Supabase. The blue progress bar moves instantly when I select an option - in case that helps? Also I have set up my Supabase to be located in us-east-1 (West Virginia)

Will, I've deployed your survey: https://pingfan.shinyapps.io/will/
My deployment went smoothly. There is no crush and I can see data updated on Supabase.
Also, although there is slow down on my local machine, this online version works out without delay.

**Comment by willking98** [comment]
Interesting that it is working well for you. I am UK based, could this be a contributing factor to the speed? I have just redeployed it and it is still crashing. In terms of the crashing - could this be related to how I am defining the password for connecting to Supabase?

**Comment by pingfan-hu** [comment]
As long as you have a password accepted by Supabase (which I believe you do), you should be fine.
All you need to do is to deliver this password to your survey project by using `surveydown::sd_set_password()`, and I believe you did so. Otherwise your local instance will fail.
Locating at UK shouldn't be a problem since it's jut a matter of connection distance (and I believe Supabase must have a server relatively closer to your location).
I simply cloned your repo, pasted my Supabase settings, and defined my password using the function and that's all.
I'm not able to debug further. I'm sorry about that.
But I would suggest you delete and reestablish a Surveydown project and try it out again.
Another suggestion from me is that it might help moving the server definition to the setup code chunk, right after `sd_setup()`. aka the `db <- sd_database()` part.
For consistency, I tested both conditions: `db <- sd_database()` stayed as it was at `server.R`, and moved to setup chunk.
- Relocated to setup chunk: https://pingfan.shinyapps.io/will/
- Old position at `server.R`: https://pingfan.shinyapps.io/will2/
Both of them worked.

**Comment by pingfan-hu** [comment]
Oh, one more thing if you are using Mac: I noticed if I installed the package and immediately attempt the deployment, it will fail or cause problems. I suggest that you reboot your computer, delete your local html file and the rsconnect folder, and deploy again. Maybe it will solve the problem.

**Comment by jmgirard** [comment]
Another thing to look out for is, if you publish to shinyapps.io, make sure you upload `.Renviron` since that is where the supabase password is actually stored.

**Comment by jhelvy** [comment]
Just leaving a note here for anyone else experiencing slow speeds, we have some ideas as to why, and it'll take some time before we can fix it, but we will be trying some strategies this week. It's entirely to do with the database as everything runs fast without a DB connection. I'm confident we can improve it, so for only speed issues just wait a little and hopefully we'll get it solved soon. 

Some of the other issues reported here are related but different, like getting an app to load on shinyapps.io. For those please start separate issues so we can address them separately.

**Comment by willking98** [comment]
> Oh, one more thing if you are using Mac: I noticed if I installed the package and immediately attempt the deployment, it will fail or cause problems. I suggest that you reboot your computer, delete your local html file and the rsconnect folder, and deploy again. Maybe it will solve the problem.

So I'm not sure whether this will be helpful or not but I have fixed both problems but don't quite know how... The survey is now running very fast online here: https://willking98.shinyapps.io/demo-choice-based-conjoint-main-1/. Responses are being saved in Supabase. 

Things I did in rough order:
1. Deleted rsconnect folder and html file
2. Restart
3. Change Supabase password
4. Reopen RStudio
5. library(surveydown), setwd(), sd_set_password("PASSWORD")
6. library(rsconnect), deployApp() [I left this function blank rather than populating it
7. Chose to host it as a new app rather than update the old one

Likely that the crashing was caused by the .renviron not being used in deployment of app. Speed wise I am not sure what the fix was.

Thank you both @pingfan-hu and @jmgirard for your help with this and congratulations again to the development team for producing such a fantastic package!

**Comment by pingfan-hu** [comment]
> > Oh, one more thing if you are using Mac: I noticed if I installed the package and immediately attempt the deployment, it will fail or cause problems. I suggest that you reboot your computer, delete your local html file and the rsconnect folder, and deploy again. Maybe it will solve the problem.
> 
> So I'm not sure whether this will be helpful or not but I have fixed both problems but don't quite know how... The survey is now running very fast online here: https://willking98.shinyapps.io/demo-choice-based-conjoint-main-1/. Responses are being saved in Supabase.
> 
> Things I did in rough order:
> 
> 1. Deleted rsconnect folder and html file
> 2. Restart
> 3. Change Supabase password
> 4. Reopen RStudio
> 5. library(surveydown), setwd(), sd_set_password("PASSWORD")
> 6. library(rsconnect), deployApp() [I left this function blank rather than populating it
> 7. Chose to host it as a new app rather than update the old one
> 
> Likely that the crashing was caused by the .renviron not being used in deployment of app. Speed wise I am not sure what the fix was.
> 
> Thank you both @pingfan-hu and @jmgirard for your help with this and congratulations again to the development team for producing such a fantastic package!

Haha I'm glad you had yours working, Will! And I appreciate your compliments. This whole idea is from my supervisor, professor John Helveston (@jhelvy). He's the core designer and main contributor to this project. For more information of our project, you can proceed to our official website: https://surveydown.org
And @jmgirard has been actively using our package and debugging the database connection problems these days. Since this connection delay issue is not completely solved, we will leave this issue open. Jeffery, please wait for our further progress regarding the connection delay.

Should you have any other problems, feel free to reply to this issue, or open up a new issue.

**Comment by pingfan-hu** [comment]
Hi @jmgirard ! Today @jhelvy has updated the database communication part and I tested with 100 questions using your template. The survey worked super fast. You can update to our newest package and have a try!

**Comment by jmgirard** [comment]
That seems to have fixed it. Nice work.

**Comment by jmgirard** [comment]
With this new update, do you recommend putting `db <- sd_database(...)` in setup or server?

**Comment by jhelvy** [comment]
Setup. We've confirmed this is correct location, and we've updated all the documentation and demo examples, e.g. [here](https://surveydown.org/survey-components#setup-code-chunk).
